pkg/instrumentation/metrics/metrics.go

```diff
@@ -15,142 +15,210 @@
+		metrics.SetProvider(nil)
+		metrics.Configure(nil)
 		return nil
 	}
 
-	log.Info("starting metrics exporter...")
-
-	g, err := metrics.NewGatherer(
-		metrics.WithNamespace(namespace),
-		metrics.WithPollInterval(reportPeriod),
-		metrics.WithMetrics(enabled, polled),
+	var (
+		ctx     = context.Background()
+		options = []metric.Option{metric.WithResource(resource)}
 	)
-	if err != nil {
-		return fmt.Errorf("failed to create metrics gatherer: %v", err)
-	}
 
-	gatherer = g
+	switch exporter {
+	case promExporter:
+		log.Info("using OpenTelemetry Prometheus exporter")
 
-	handlerOpts := promhttp.HandlerOpts{
-		ErrorLog:      log,
-		ErrorHandling: promhttp.ContinueOnError,
-	}
-	m.Handle("/metrics", promhttp.HandlerFor(g, handlerOpts))
+		// To enable/disable 'standard' OpenTelemetry or runtime-provided
+		// metrics we either use the default prometheus registerer (enabled)
+		// or one-off custom one (disabled).
+		registry := prometheus.DefaultRegisterer
+		if !metrics.IsEnabled("standard", "") {
+			registry = prometheus.NewRegistry()
+		}
+		gatherer := registry.(prometheus.Gatherer)
+
+		exp, err := otelprom.New(
+			otelprom.WithNamespace(namespace),
+			otelprom.WithRegisterer(registry),
+			otelprom.WithoutScopeInfo(),
+			otelprom.WithoutTargetInfo(),
+		)
+		if err != nil {
+			return fmt.Errorf("failed to create OpenTelemetry Prometheus exporter: %w", err)
+		}
 
-	mux = m
+		options = append(options, metric.WithReader(exp))
 
-	return nil
-}
+		handlerOpts := promhttp.HandlerOpts{
+			ErrorHandling: promhttp.ContinueOnError,
+		}
+		m.Handle("/metrics", promhttp.HandlerFor(gatherer, handlerOpts))
 
-func Stop() {
-	if mux == nil {
-		return
+	case httpExporter, "otel-http":
+		log.Info("using OpenTelemetry HTTP exporter")
+
+		exp, err := otlpmetrichttp.New(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to create OpenTelemetry HTTP exporter: %w", err)
+		}
+
+		options = append(options,
+			metric.WithReader(
+				metric.NewPeriodicReader(exp, metric.WithInterval(reportPeriod)),
+			),
+		)
+
+	case grpcExporter, "otel-grpc":
+		log.Info("using OpenTelemetry gRPC exporter")
+
+		exp, err := otlpmetricgrpc.New(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to create OpenTelemetry gRPC exporter: %w", err)
+		}
+
+		options = append(options,
+			metric.WithReader(
+				metric.NewPeriodicReader(exp, metric.WithInterval(reportPeriod)),
+			),
+		)
 	}
 
-	mux.Unregister("/metrics")
-	mux = nil
-	gatherer.Stop()
-	gatherer = nil
-}
+	log.Info("starting metrics exporter...")
 
-func Block() *MetricsBlock {
-	return newMetricsBlock(gatherer)
-}
+	provider = metric.NewMeterProvider(options...)
+	metrics.SetProvider(provider)
+
+	mux = m
 
-type MetricsBlock struct {
-	g *metrics.Gatherer
+	return nil
 }
 
-func newMetricsBlock(g *metrics.Gatherer) *MetricsBlock {
-	if g == nil {
-		return nil
+// Stop metrics collection and exporting.
+func Stop() {
+	if provider != nil {
```

Comment:
Without really knowing internals here, would stopping be safer if the provider would be first unregistered from the mux, and shutdown only after that? (Mentally trying to avoid situation where the provider would get a request from the mux while or after being shut down.)

Reply:
Changed ordering of Unregister and shutdown.